addons/l10n_ro_efactura/controllers/main.py

```diff
@@ -29,28 +32,42 @@ def callback(self, company_id, **kw):
         # Without certificate, ANAF won't give any access key in the callback URL's "code" parameter
         if not access_key:
             error_message = _("Access key not found. Please try again.\nResponse: %s", kw)
-            company.l10n_ro_edi_oauth_error = error_message
-            company.env.cr.commit()
+            company._l10n_ro_edi_log_message(error_message)
+            raise UserError(error_message)
+
+        try:
+            response = requests.post(
+                url='https://logincert.anaf.ro/anaf-oauth2/v1/token',
+                data={
+                    'grant_type': 'authorization_code',
+                    'client_id': company.l10n_ro_edi_client_id,
+                    'client_secret': company.l10n_ro_edi_client_secret,
+                    'code': access_key,
+                    'access_key': access_key,
+                    'redirect_uri': company.l10n_ro_edi_callback_url,
+                    'token_content_type': 'jwt',
+                },
+                headers={
+                    'content-type': 'application/x-www-form-urlencoded',
+                    'accept': 'application/json',
+                    'user-agent': 'PostmanRuntime/7.29.2',
+                },
+                timeout=10,
+            )
+        except Exception as e:
+            error_message = f"Request to https://logincert.anaf.ro/anaf-oauth2/v1/token failed: {str(e)}\n" \
+                            f"Received access_key: {access_key}"
+            company._l10n_ro_edi_log_message(error_message)
+            raise UserError(error_message)
+
+        try:
+            response_json = response.json()
+            company._l10n_ro_edi_process_token_response(response_json)
+        except Exception as e:
+            error_message = f"Error when processing the response: {str(e)}\n" \
+                            f"Received access_key: {access_key}\n\n" \
+                            f"{dump.dump_all(response).decode('latin-1')}"
```

Comment:
I have my reasons... Let's discuss this in person tomorrow :ok_hand:

Reply:
@aboo-odoo @yosuanicolaus please, at least put what was decided here. We weren't there in person with you :smile:

Reply:
@rugo-odoo discussion TLDR: 

- `dump_all` is a bad idea because we might log sensitive information. Turns out `response.text` already covers everything we'll need. https://github.com/odoo/odoo/pull/187708/files#diff-7aeb164f7a36076d64e777aeec6d719fa981fb3452ad5d3db9da221a0d4de02bR69-R72 ;)
- `binascii.Error` is for handling if somehow `base64.b64decode(payload, altchars=b'-_', validate=True)` fails
- since it's raised in UserError, the new error messages should be translatable
- we currently don't know where it might raise an error, so just to be safe, we have the `except Exception` at the end to catch any kind of error from this whole auth process